src/Service/RedirectedURLService.php

```diff
@@ -0,0 +1,133 @@
+        $getVars = $this->arrayToLowercase($request->getVars());
+
+        // Find all the RedirectedURL objects where the base URL matches.
+        // Assumes the base url has no trailing slash.
+        $SQL_base = Convert::raw2sql(rtrim($base, '/'));
+
+        $potentials = RedirectedURL::get()->filter(['FromBase' => '/' . $SQL_base])->sort('FromQuerystring DESC');
+        $listPotentials = new ArrayList();
+        foreach ($potentials as $potential) {
+            $listPotentials->push($potential);
+        }
+
+        // Find any matching FromBase elements terminating in a wildcard /*
+        $baseparts = explode('/', $base);
+        for ($pos = count($baseparts) - 1; $pos >= 0; $pos--) {
+            $basestr = implode('/', array_slice($baseparts, 0, $pos));
+            $basepart = Convert::raw2sql($basestr . '/*');
+            $basepots = RedirectedURL::get()->filter(['FromBase' => '/' . $basepart])->sort('FromQuerystring DESC');
+            foreach ($basepots as $basepot) {
+                // If the To URL ends in a wildcard /*, append the remaining request URL elements
+                if ($basepot->RedirectionType === 'External' && substr($basepot->To, -2) === '/*') {
+                    $basepot->To = substr($basepot->To, 0, -2) . substr($base, strlen($basestr));
+                }
+                $listPotentials->push($basepot);
+            }
+        }
+
+        $matched = null;
+
+        // Then check the get vars, ignoring any additional get vars that
+        // this URL may have
+        if ($listPotentials) {
+            foreach ($listPotentials as $potential) {
+                $allVarsMatch = true;
+
+                if ($potential->FromQuerystring) {
+                    $reqVars = array();
+                    parse_str($potential->FromQuerystring, $reqVars);
+
+                    foreach ($reqVars as $k => $v) {
+                        if (!$v) {
+                            continue;
+                        }
+
+                        if (!isset($getVars[$k]) || $v != $getVars[$k]) {
+                            $allVarsMatch = false;
+                            break;
+                        }
+                    }
+                }
+
+                if ($allVarsMatch) {
+                    $matched = $potential;
+                    break;
+                }
+            }
+        }
+
+        // If we found a match, we return it - otherwise we return null to indicate that no match was found
+        return $matched;
+    }
+
+    public function getResponse(RedirectedURL $redirect): HTTPResponse
+    {
+        $response = HTTPResponse::create()
+            ->redirect(Director::absoluteURL($redirect->Link()), $this->getRedirectCode($redirect));
+
+        return $response;
+    }
+
+    /**
+     * Converts an array of key value pairs to lowercase
+     *
+     * @param array $vars key value pairs
+     * @return array
+     */
+    protected function arrayToLowercase($vars)
```

Comment:
Unless we explicitly expect people to extend this class and call this method, I would make it private to minimise our API surface.